app/Policies/SongPolicy.php

```diff
@@ -28,7 +28,7 @@ public function viewAny(User $user)
      * @param  \App\Song  $song
      * @return mixed
      */
-    public function view(User $user, Song $song)
+    public function view(?User $user, Song $song)
     {
         return $song->is_active || $user->is($song->artist->user);
     }
```

Comment:
if user is optional you need something like:

```php
return $song->is_active || optional($user)->is($song->artist->user);
```

or assuming that a song artist user is always there flip it:

```php
return $song->is_active || $song->artist->user->is($user);
```

Reply:
Good catch! And it should actually be (we had forgotten the `album` part):

```php
return $song->is_active || $song->album->artist->user->is($user);
```

In fact, now that I think about this a bit more, it also seems prudent to add tests to ensure that:

1. When a Song is inactive and a User is not logged-in, access is denied.
2. When a Song is inactive and a User is logged-in, but the User doesn't own the Song, access is denied.
3. When a Song is inactive, but the User owns the Song, access is allowed.
4. When a Song is active, but its parent Album is inactive, and a User is not logged-in, access is denied. (The logic here is that an Album's active/inactive status "overrides" the individual Song's status for non-owners; the owner has access regardless of both the Album and Song's statuses.)

These rules would change the line in question as such, if I'm not mistaken:

```php
return ($song->album->is_active && $song->is_active)
            || $song->album->artist->user->is($user);
```

Reply:
I fixed this and added several new tests to verify these requirements. Take a look whenever you get a chance!